sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2796,7 +2796,31 @@ checkDevSupportDeviceRequirements(const device &Dev,
       Dims++;
     }
 
-    if (NDRDesc.Dims != 0 && NDRDesc.Dims != static_cast<size_t>(Dims))
+    int NumDimSpecified = 0;
+    if (WorkGroupNumDim) {
+      // We know the dimensions have been padded to 3, make sure that the pad
+      // value is always set to 1 and that the non-padded dimension matches the
+      // ones specified in the ND range.
+      NumDimSpecified = static_cast<int>(
+          DeviceBinaryProperty(*(WorkGroupNumDim.value())).asUint32());
+      if (NDRDesc.Dims != 0 &&
+          NDRDesc.Dims != static_cast<size_t>(NumDimSpecified))
+        return sycl::exception(
+            sycl::errc::nd_range,
+            "The local size dimension of submitted nd_range doesn't match the "
+            "required work-group size dimension");
```

Comment:
Maybe we can make a single if-statement for the exception? I'd be okay with making the conditions into bool declarations to make a single if-statement more readable.

Reply:
That's a good idea, I've attempted it here: https://github.com/intel/llvm/pull/13600/commits/78941a098758e2e16bf3fbb8280d00be3f201eda